src/jit/lclvars.cpp

```diff
@@ -235,6 +235,52 @@ void Compiler::lvaInitTypeRef()
 
     lvaInitArgs(&varDscInfo);
 
+#if FEATURE_FASTTAILCALL
+
+    //-------------------------------------------------------------------------
+    // Calculate the argument register usage.
+    //
+    // This will later be used for fastTailCall determination
+    //-------------------------------------------------------------------------
+
+    unsigned argRegCount      = 0;
+    unsigned floatingRegCount = 0;
+    unsigned stackArgCount    = 0;
+    size_t   stackSize        = 0;
+    unsigned compArgCount     = info.compArgsCount;
+
+    auto incrementRegCount = [&floatingRegCount, &argRegCount](LclVarDsc* varDsc) {
+        varDsc->IsFloatRegType() ? ++floatingRegCount : ++argRegCount;
+    };
+
+    unsigned   argNum;
+    LclVarDsc* curDsc;
+
+    for (curDsc = lvaTable, argNum = 0; argNum < varDscInfo.varNum; argNum++, curDsc++)
+    {
+        if (curDsc->lvIsRegArg)
+        {
+            incrementRegCount(curDsc);
+
+#if FEATURE_MULTIREG_ARGS
+            if (curDsc->lvOtherArgReg != REG_NA)
+            {
+                incrementRegCount(curDsc);
+            }
+#endif // FEATURE_MULTIREG_ARGS
+        }
+        else if (varTypeIsStruct(curDsc))
+        {
+            stackSize += curDsc->lvSize();
+        }
+        else
+        {
+            stackSize += TARGET_POINTER_SIZE;
+        }
+    }
```

Comment:
Yes, I think it is risky to have this logic duplicated. This same logic exists, in a different form, in `fgMorphArgs()`,yet there's no clear way to ensure consistency between these two pieces of code that need to understand how the args are mapped to regs/stack.

---
In reply to: [124417850](https://github.com/dotnet/coreclr/pull/12489#discussion_r124417850) [](ancestors = 124417850)